src/repl/scala/tools/nsc/interpreter/Scripted.scala

```diff
@@ -0,0 +1,292 @@
+
+    // $ctx.x retrieves the attribute x
+    def selectDynamic(field: String): Object = context.getAttribute(field)
+
+    // $ctx.x = v
+    def updateDynamic(field: String)(value: Object) = context.setAttribute(field, value, ENGINE_SCOPE)
+
+    // set the context for dynamic resolution and run the body
+    def apply[A](context: ScriptContext)(body: => A): A = {
+      val saved = this.context
+      this.context = context
+      try body
+      finally this.context = saved
+    }
+  }
+
+  // Defines attributes available both for compilation and evaluation
+  val dynamicContext = new DynamicContext
+
+  // the underlying interpreter
+  val intp = new IMain(settings, out) {
+    import global.{ Name, TermName }
+
+    // dynamic context bound under this name
+    final val ctx = "$ctx"
+
+    /* Modify the template to snag definitions from dynamic context.
+     * So object $iw { x + 42 } becomes object $iw { def x = $ctx.x ; x + 42 }
+     */
+    override protected def importsCode(wanted: Set[Name], wrapper: Request#Wrapper, definesClass: Boolean, generousImports: Boolean) = {
+
+      // cull references that can be satisfied from the current dynamic context
+      val contextual = wanted & contextNames
+
+      if (contextual.nonEmpty) {
+        val neededContext = (wanted &~ contextual) + TermName(ctx)
+        val ComputedImports(header, preamble, trailer, path) = super.importsCode(neededContext, wrapper, definesClass, generousImports)
+        val adjusted = contextual.map(n =>
+            s"""def ${n.decodedName} = $ctx.${n.decodedName}
+                def ${n.decodedName}_=(x: Object) = $ctx.${n.decodedName} = x"""
+          ).mkString(preamble, "\n", "\n")
+        ComputedImports(header, adjusted, trailer, path)
+      }
+      else super.importsCode(wanted, wrapper, definesClass, generousImports)
+    }
+
+    // names available in current dynamic context
+    def contextNames: Set[Name] = (
+      for {
+        scope   <- dynamicContext.context.getScopes.asScala
+        binding <- Option(dynamicContext.context.getBindings(scope)) map (_.asScala) getOrElse Nil
+        k = binding._1
+      } yield (TermName(k) : Name)
+    ).to[Set]
+  }
+
+  private var bound = false
+  private var code = ""
+
+  /* All scripts are compiled. The supplied context defines what references
+   * not in REPL history are allowed, though a different context may be
+   * supplied for evaluation of a compiled script.
+   */
+  def compile(script: String, context: ScriptContext): CompiledScript = {
+    // TODO a mechanism for post-init actions
+    if (!bound) {
+      //intp quietBind ("engine" -> this.asInstanceOf[ScriptEngine])
+      //intp quietBind NamedParam[DynamicContext]("$ctx", dynamicContext)
+      intp quietBind NamedParamClass(intp.ctx, "scala.tools.nsc.interpreter.Scripted#DynamicContext", dynamicContext)
+      intp quietBind NamedParam[IMain]("$intp", intp)(StdReplTags.tagOfIMain, classTag[IMain])
```

Comment:
Regarding this line, what I intended was to make the interpreter available as a ScriptEngine, that is, it should be "this" and not "intp" (and the name was "engine", without $, it would be cool if it could remain like this).